DotvvmProperty aliasing

cafour opened the pull request:
Implements support for `DotvvmProperty` aliasing (part of #420) so that dotproperty renames and deprecation are possible (and thus issues like #1091 can be implemented).

cafour commented:
* [ ] Allow aliasing of attached properties from two different classes.
* [ ] Add metadata for the VS extension.
* [ ] Register aliases in two phases so as to avoid having to order the static fields.
* [ ] Target `v4-main`.

exyi commented:
I'd personally prefer to do the aliasing at compile time and disable use of the property completely at runtime. This would have the following properties compared to this approach:

* Usage of both the old and new property will be caught by the compiler automatically.
* C# Code using the property will have to be migrated immediately
   - IMHO, this is advantage, since the GetValue / SetValue overloads are bit dubious, we don't use them everywhere (for example, when iterating control.Properties)
   - On the other hand, changing the name will be a "more severe" breaking change
* The GetValue / SetValue overloads are not respected in compile-time styles, so it could lead to inconsistencies.. Rewriting the properties at compile-time before the styles are applied would solve the problem.

MichalTichy commented:
@exyi 
Can you elaborate on:
- Usage of both the old and new property will be caught by the compiler automatically.

When using deprecated prop with alias I would want to be able to keep using it without any compilation errors (warnings would be really cool).

Are you suggesting that we should throw some exception witch will tell you what prop to use instead?

This behavior is not something I would prefer, instead I would just like to get some sort of notification that I should not use this prop (and extension should not suggest this prop when writing new code).
This would allow me to "fix" my code over time.

cafour commented:
@exyi
* How would you declare this compile-time aliasing? By adding `Aliases` to `MarkupOptionsAttribute` or in different way?
* Personally, I think that just the compile-time approach would be too severe. If we were to go through the renames I proposed in #1091, we'd make potentially dozens of aliases. It'd be far less intrusive to deprecate the original properties in one major release and remove them in the one after that.

> Usage of both the old and new property will be caught by the compiler automatically.

That's true for the runtime approach as well, isn't it? You can still check for the presence of an `ObsoleteAttribute`.

> The GetValue / SetValue overloads are not respected in compile-time styles, so it could lead to inconsistencies...

Honestly, I forgot about server-side styles. :D However, looking over their implementation, they seem they could be made aware of the aliasing (in its present implementation) quite easily.

exyi commented:
> Usage of both the old and new property will be caught by the compiler automatically.

Sorry, I probably didn't manage to explain what I mean. Let's say we rename `Checked` to `IsChecked`, then I'll be able to write `<dot:CheckBox Checked={value: X} IsChecked={value: Y}` and will get no error (at least with the current implementation). You should be allowed to use the old property and of course the new property, just not both on the same control.

exyi commented:
> Are you suggesting that we should throw some exception witch will tell you what prop to use instead?

Yes, but only if you do that at runtime (i.e. not in dothtml, but in C#). Since most people use controls from dothtml, I don't think this will create many problems.

MichalTichy commented:
> > Usage of both the old and new property will be caught by the compiler automatically.
> 
> Sorry, I probably didn't manage to explain what I mean. Let's say we rename `Checked` to `IsChecked`, then I'll be able to write `<dot:CheckBox Checked={value: X} IsChecked={value: Y}` and will get no error (at least with the current implementation). You should be allowed to use the old property and of course the new property, just not both on the same control.

Yeah using both at the same time should throw exception.
But I'm still not sold on throwing exception when anyone uses the old prop.

cafour commented:
@exyi I implemented some of the changes we discussed last Friday: I added an `AliasingVisitor` and an `ObsoletionVisitor` that handle compile-time aliasing and propagation of the `Obsolete` attribute into Dothtml respectively.

However, I also found a little issue with the `Obsolete` attribute. We'd like it so that in C#, using an alias mean a compilation error but using the same alias in dothtml should create a mere warning (unless the aliased property is set as well). The issue is that the propagation of deprecation warnings and errors into Dothtml is and IMHO should be a feature separate from property aliasing. There are after all more reasons to deprecate a property than just aliasing. Therefore if there is an `Obsolete` attribute with `IsError=true` there should always be an error in Dothtml as well.

If aliasing should work the way we wanted it, we'd then have to explicitly handle the special case where an alias should generate an error in C# but a warning in Dothtml, which is not very intuitive. Is there any other catch other than the server-side styles that would prevent the current implementation from aliasing properly both in compile-time and runtime?

exyi reviewed (commented):
There is still a few problems:

* if I'd set the aliased property in server-side styles, it will not be rewritten
* if I'll ask for the aliased property in server side styles, it will always be unset


I guess these are kinda acceptable glitches, just please revert the redundant changes, I don't want to deal with unnecessary merge conflicts (in PropertyGroupDescriptor and maybe some in DotvvmProperty, I have 2 branching changing those)

cafour commented on src/DotVVM.Framework/DependencyInjection/DotVVMServiceCollectionExtensions.cs:
```diff
@@ -93,6 +93,8 @@ public static IServiceCollection RegisterDotVVMServices(IServiceCollection servi
 
             services.ConfigureWithServices<ViewCompilerConfiguration>((o, s) => {
                 var controlResolver = s.GetRequiredService<IControlResolver>();
+                o.TreeVisitors.Add(() => ActivatorUtilities.CreateInstance<ObsoletionVisitor>(s));
```
"Obsoletion" and "Obsolescence" are synonyms. I agree that the former looks a bit weird and books seem to agree with you. [1] However, "Obsoletion" is shorter and much more frequent in `dotnet/runtime` [2], so I'd rather leave it as is.

[1]: https://books.google.com/ngrams/graph?content=obsoletion%2Cobsolescence
[2]: `git grep obsolescence` vs `git grep obsoletion`

exyi commented:
Think about it a bit more, I'd be much more happy to accept the weird case of error/warning in dothtml and disallow usage from C# completely than having server-side styles glitched.

cafour commented:
@exyi I'm fine with disallowing usage of the aliased property in C# completely, just not at the expense of the `ObsoleteAttribute` propagation feature: We'd have to assume that the obsolete attribute on a property alias should never create an error in dothtml, despite having `IsError = true` and hardcode this behavior into `ObsoletionVisitor`.

Maybe there is another solution that I'm not seeing. If you have an idea, I'd be happy to implement it.

cafour commented on src/DotVVM.Framework/Compilation/ControlTree/DotvvmPropertyGroup.cs:
```diff
@@ -58,32 +63,49 @@ protected DotvvmPropertyGroup (PropertyInfo propertyInfo, PrefixArray prefixes,
+            FieldInfo descriptorField,
+            ICustomAttributeProvider attributeProvider,
+            string name,
+            object? defaultValue)
```
To add the `ObsoleteAttribute` property. I limited the unnecessary changes, though.
